ilastik/applets/wsdt/opWsdt.py

```diff
@@ -1,13 +1,141 @@
+      halo: portion of each block to discard after processing for smoother boundary regions
+        if not specified: 10 voxels around the block in each direction
+      max_workers: if not specified or None, will use number of workers in the global Requests threadpool
+
+    """
+
+    logger.info(f"blockwise watershed with {max_workers} threads.")
+    shape = data.shape
+    ndim = len(shape)
+
+    assert ndim in [2, 3], "Watershed segmentor will only work on 2D and 3D data"
+
+    base_block = 512 if ndim == 2 else 128
+    # check for None arguments and set to default values
+    block_shape = (base_block,) * ndim if block_shape is None else block_shape
+    # NOTE nifty expects a list for the halo parameter although a tuple would be more logical
+    halo = [10] * ndim if halo is None else halo
+    blocking = get_blocking(data, block_shape, roi=None)
+
+    max_workers = max_workers = max(1, Request.global_thread_pool.num_workers) if max_workers is None else max_workers
+
+    n_blocks = blocking.numberOfBlocks
+
+    labels = np.zeros(shape, dtype="uint32")
+
+    # watershed for a single block
+    def ws_block(block_index):
+        nonlocal labels
+
+        # get the block with halo and the slicings corresponding to
+        # the block with halo, the block without halo and the
+        # block without halo in loocal coordinates
+        block = blocking.getBlockWithHalo(blockIndex=block_index, halo=halo)
+        inner_slicing = roiToSlice(block.innerBlock.begin, block.innerBlock.end)
+        outer_slicing = roiToSlice(block.outerBlock.begin, block.outerBlock.end)
+        inner_local_slicing = roiToSlice(block.innerBlockLocal.begin, block.innerBlockLocal.end)
+
+        with Timer() as btimer:
+            # write watershed result to the label array
+            ws_outer, _ = distance_transform_watershed(
+                data[outer_slicing],
+                threshold,
+                sigma_seeds,
+                sigma_weights,
+                minsize,
+                alpha,
+                pixel_pitch,
+                non_max_suppression,
+            )
+
+        logger.debug(
+            f"\tprocessing block {block_index} {block.outerBlock.begin}-{block.outerBlock.end} took {btimer.seconds()}"
+        )
+
+        ws_inner = vigra.analysis.labelMultiArray(ws_outer[inner_local_slicing])
+
+        labels[inner_slicing] = ws_inner
+        # return the max-id for this block, that will be used as offset
+        return ws_inner.max()
+
+    with Timer() as wstimer:
+        # run the watershed blocks in parallel
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            tasks = [executor.submit(ws_block, block_index) for block_index in range(n_blocks)]
+            offsets = np.array([t.result() for t in tasks], dtype="int64")  # TODO uint32
+
+    logger.info(f"parallel ws took {wstimer.seconds()} s")
+    # compute the block offsets and the max id
+    last_max_id = offsets[-1]
```

Comment:
If I'm not mistaken, we could simplify offset computation like this:

```python
def ws_block(...):
    # Return the next possible ID for this block.
    return ws_inner.max() + 1

...

numpy.cumsum(offsets, out=offsets)

def add_offset_block(block_index):
    ...
    labels[block] += offsets[block_index - 1]

with ThreadPoolExecutor(max_workers=max_workers) as executor:
    # Labels in the first block don't need to change.
    [executor.submit(add_offset_block, block_index) for block_index in range(1, n_blocks)]

return labels, offsets[-1]
```

Reply:
hmm, not exactly. In the current code, 1 is added to the first block, this would not be the case with this code (as you acknowledged indirectly with the comment that first block doesn't need any changing). While this is not necessary for multicut, it is for carving. And at some point I'd like to reuse this function there... I'll update it with a comment.

Reply:
Hmm, I see. In this case, do we even need offsets as numpy arrays? We already collect them sequentially from `Executor.map`, so we might as well prefix-sum them right there.
```python
offsets = itertools.accumulate(executor.map(ws_block, range(n_blocks)), initial=1)
```
Obviously, in this case we return `int(ws_inner.max())`, and index `offsets` with just `block_index` without `-1`.

At least it would be nice to avoid explicit `max_id/last_max_id` variables because it's normal for prefix sum idioms to have some starting element (1 in our case) and have something like `prefix_sum[-1] == total_max`.

Reply:
initial was sadly only added in  3.8...

Reply:
well, in the end:

https://github.com/ukoethe/vigra/blob/9b514fa00a136f5fd81bb57ee9f6293c333ffc1f/include/vigra/multi_watersheds.hxx#L456-L457

vigra already returns the labels starting at 1... I added a trivial test to confirm this.